File: code/ex362.java
// O(1)

class HitCounter {
    private int[] hit;    // Record the number of hits on that timestamp
    private int[] ts;    // Record the actual timestamp on that position

    /** Initialize your data structure here. */
    public HitCounter() {
        hit = new int[300];
        ts = new int[300];
    }
    
    /** Record a hit.
        @param timestamp - The current timestamp (in seconds granularity). */
    public void hit(int timestamp) {
        if (ts[timestamp % 300] == timestamp) {
            hit[timestamp % 300]++;
        } else {
            ts[timestamp % 300] = timestamp;
            hit[timestamp % 300] = 1;
        }
    }
    
    /** Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity). */
    public int getHits(int timestamp) {
        int total = 0;
        for (int i = 0; i < 300; i++)
            if (timestamp - ts[i] < 300)
                total += hit[i];
        return total;
    }
}

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter obj = new HitCounter();
 * obj.hit(timestamp);
 * int param_2 = obj.getHits(timestamp);
 */

File: code/ex362.cpp
class HitCounter {
    deque<pair<int, int>> dque;
    int inque = 0;
public:
    /** Initialize your data structure here. */
    HitCounter() {
        
    }
    
    /** Record a hit.
        @param timestamp - The current timestamp (in seconds granularity). */
    void hit(int timestamp) {
        if (!dque.empty() && dque.back().first == timestamp) {
            pair<int, int> out = dque.back();    dque.pop_back();
            dque.push_back({timestamp, out.second + 1});
        } else {
            dque.push_back({timestamp, 1});
        }
        inque++;
    }
    
    /** Return the number of hits in the past 5 minutes.
        @param timestamp - The current timestamp (in seconds granularity). */
    int getHits(int timestamp) {
        while (!dque.empty() && timestamp - dque.front().first >= 5 * 60) {
            inque -= dque.front().second;
            dque.pop_front();
        }
        return inque;
    }
};

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter* obj = new HitCounter();
 * obj->hit(timestamp);
 * int param_2 = obj->getHits(timestamp);
 */